### DISCUSSION TITLE
Embedding inline images in emails broken when upgrading from Laravel 9 to 11

### DISCUSSION THREAD

**Discussion by phjl** [laravel/framework#51164] [Q&A]
I noticed  after updating from Laravel 9 to 11 that embedded email images stopped showing and the code below started returning the attachment filenames like `JipwOYRW4e` instead of the Content-ID like `f150f0a9b59076bc9768a5910cce5d7b@symfony`

```php
$message->embed(resource_path('assets/images/email/myimg.png'))
```

If I override `Illuminate/Mail/Message::embed()` to return the Content ID instead like below the images end up being displayed like before again:

```php
$cid = Str::random(10);
$dataPart = (new DataPart(new File($file), $cid))->asInline();
$this->message->addPart(
    $dataPart
);

return "cid:" . $dataPart->getContentId();
```

Obviously I'd rather not override framework classes though so what's a better way of fixing this?

**Comment by zjean** [comment]
@phjl How did you override the Message class and use that in your mailable?

**Reply by phjl** [reply]
Hi @zjean! 
I just edited it directly in my vendor directory for testing purposes.

The solution I ended up going with for now though is registering a listener for the `Iluminate\Mail\Events\MessageSending` event and just having it rewrite the cid parts to use the content ids instead of the filenames:
```php
public function handle(MessageSending $event)
{
    $body = $event->message->getBody();
    $mediaType = $body->getMediaType();
    $mediaSubtype = $body->getMediaSubtype();

    if (
        ($mediaType === 'text' && $mediaSubtype === 'html') ||
        ($mediaType === 'multipart')
    ) {
        $message = $event->message;
        $body = $event->message->getHtmlBody();

        $dataParts = $event->message->getAttachments();
        foreach ($dataParts as $dataPart) {
            if ($dataPart->getDisposition() == "inline") {
                $body = str_replace("cid:" . $dataPart->getFileName(), "cid:" . $dataPart->getContentId(), $body);
            }
        }

        $message->html($body);
    }
}
```
